File: tests/python/realworld_open_redirect.py
"""
Real-world open redirect test cases for Kunlun-M scanner.
Simulates a Flask application with login/redirect flows.
"""

from urllib.parse import urlparse, urljoin
from flask import Blueprint, request, redirect, jsonify, url_for

auth_bp = Blueprint('auth', __name__)

# Whitelist of allowed redirect destinations
ALLOWED_REDIRECT_DOMAINS = [
    'app.example.com',
    'dashboard.example.com',
    'docs.example.com',
]


# ---------------------------------------------------------------------------
# VULN 1: redirect with user-controlled URL from query parameter
# ---------------------------------------------------------------------------
@auth_bp.route('/login', methods=['POST'])
def login():
    """Handle user login and redirect to next page."""
    username = request.form.get('username', '')
    password = request.form.get('password', '')
    next_url = request.args.get('next', '/')
    # VULN: User-controlled URL used directly in redirect
    # Attacker crafts link: /login?next=https://evil.com/phishing
    # After successful login, user is redirected to attacker's phishing page
    return redirect(next_url)


# ---------------------------------------------------------------------------
# VULN 2: redirect with user-controlled URL from POST body
# ---------------------------------------------------------------------------
@auth_bp.route('/auth/callback', methods=['POST'])
def oauth_callback():
    """OAuth callback with redirect URL."""
    redirect_uri = request.form.get('redirect_uri', '/')
    # VULN: User-controlled redirect_uri passed directly to redirect
    # Attacker posts: redirect_uri=https://attacker.com/steal-token
    return redirect(redirect_uri)


# ---------------------------------------------------------------------------
# VULN 3: 302 redirect with user-controlled Location header
# ---------------------------------------------------------------------------
@auth_bp.route('/auth/continue', methods=['GET'])
def continue_auth():
    """Continue authentication flow to external provider."""
    target = request.args.get('target', '')
    # VULN: User-controlled target used in redirect
    # Attacker uses: /auth/continue?target=https://evil.com
    from flask import make_response
    resp = make_response('', 302)
    resp.headers['Location'] = target
    return resp


# ---------------------------------------------------------------------------
# SAFE 1: Relative redirect only (URL starts with /)
# ---------------------------------------------------------------------------
@auth_bp.route('/auth/logout', methods=['POST'])
def logout():
    """Handle user logout with safe redirect."""
    next_page = request.form.get('next', '/')
    # SAFE: Ensure redirect is relative (starts with /, no protocol)
    if not next_page.startswith('/'):
        next_page = '/'
    return redirect(next_page)


# ---------------------------------------------------------------------------
# SAFE 2: Domain whitelist validation before redirect
# ---------------------------------------------------------------------------
@auth_bp.route('/auth/sso/return', methods=['GET'])
def sso_return():
    """SSO return endpoint with redirect validation."""
    return_to = request.args.get('return_to', '/dashboard')
    # SAFE: Validate redirect destination against whitelist
    parsed = urlparse(return_to)
    if parsed.hostname and parsed.hostname not in ALLOWED_REDIRECT_DOMAINS:
        return_to = '/dashboard'
    return redirect(return_to)

File: tests/php/realworld_open_redirect.php
<?php
/**
 * Real-world Open Redirect test file for Kunlun-M scanner.
 * Simulates login/oauth callback flow with redirect functionality.
 */

class AuthController {
    private $allowedDomains = [
        'example.com',
        'www.example.com',
        'app.example.com',
        'dashboard.example.com',
    ];
    private $trustedRedirects = [
        '/dashboard',
        '/profile',
        '/settings',
        '/home',
        '/account',
    ];

    /**
     * VULN: Open redirect via $_GET['redirect'] in header Location.
     * Attacker can use: ?redirect=https://evil.com/phishing
     */
    public function loginRedirect() {
        $redirect = $_GET['redirect'] ?? '/dashboard';

        // Simulated login check
        $_SESSION['user_id'] = 1;

        // Direct redirect without validation
        header("Location: " . $redirect);
        exit;
        // Taint source: $_GET['redirect'] -> Sink: header("Location:")
    }

    /**
     * VULN: Open redirect via $_POST['next'] in header after form submission.
     * Attacker can submit form with next=https://attacker.com/steal-credentials
     */
    public function postLoginRedirect() {
        $username = $_POST['username'] ?? '';
        $password = $_POST['password'] ?? '';
        $next     = $_POST['next'] ?? '/dashboard';

        // Simplified auth check
        if (empty($username) || empty($password)) {
            http_response_code(401);
            echo json_encode(['error' => 'Missing credentials']);
            return;
        }

        // Assume auth succeeds
        $_SESSION['user'] = $username;

        header("Location: " . $next);
        exit;
        // Taint source: $_POST['next'] -> Sink: header("Location:")
    }

    /**
     * VULN: Open redirect via $_REQUEST['return_to'] after OAuth flow.
     * Uses urlencode then decode but no URL validation.
     */
    public function oauthCallback() {
        $returnTo = $_REQUEST['return_to'] ?? '/dashboard';

        // The URL was "encoded" but we decode it and trust it
        $decodedReturnTo = urldecode($returnTo);

        // Simulate setting OAuth token
        $_SESSION['oauth_token'] = 'fake_token_' . bin2hex(random_bytes(8));

        header("Location: " . $decodedReturnTo);
        exit;
        // Taint source: $_REQUEST['return_to'] -> Sink: header("Location:")
    }

    /**
     * SAFE: Whitelist-based redirect - only allow known internal paths.
     */
    public function safeLoginRedirect() {
        $redirect = $_GET['redirect'] ?? '/dashboard';

        // Strict whitelist of allowed paths
        if (!in_array($redirect, $this->trustedRedirects, true)) {
            $redirect = '/dashboard';
        }

        header("Location: " . $redirect);
        exit;
        // Safe: whitelist prevents external URL redirects
    }

    /**
     * SAFE: Validate that redirect URL is a relative path starting with /.
     */
    public function safePostLoginRedirect() {
        $username = $_POST['username'] ?? '';
        $password = $_POST['password'] ?? '';
        $next     = $_POST['next'] ?? '/dashboard';

        if (empty($username) || empty($password)) {
            http_response_code(401);
            echo json_encode(['error' => 'Missing credentials']);
            return;
        }

        // Only allow relative paths starting with /
        if (!preg_match('#^/[a-zA-Z0-9\-_/.]*$#', $next)) {
            $next = '/dashboard';
        }

        // Extra safety: block protocol schemes
        if (stripos($next, '://') !== false || stripos($next, '\\\\') !== false) {
            $next = '/dashboard';
        }

        $_SESSION['user'] = $username;
        header("Location: " . $next);
        exit;
        // Safe: regex validation + protocol scheme check
    }

    /**
     * SAFE: Parse URL and validate domain against whitelist.
     */
    public function safeOauthCallback() {
        $returnTo = $_REQUEST['return_to'] ?? '/dashboard';

        // If it's a relative URL (starts with /), it's safe
        if (preg_match('#^/([a-zA-Z0-9\-_/.]*)$#', $returnTo)) {
            header("Location: " . $returnTo);
            exit;
        }

        // If absolute URL, validate domain
        $parsed = parse_url($returnTo);
        if ($parsed && isset($parsed['host'])) {
            if (in_array($parsed['host'], $this->allowedDomains, true)) {
                header("Location: " . $returnTo);
                exit;
            }
        }

        // Fallback to dashboard
        header("Location: /dashboard");
        exit;
        // Safe: relative path allowed + absolute URL domain whitelist
    }
}

// --- Routing ---
if (php_sapi_name() !== 'cli') {
    if (session_status() === PHP_SESSION_NONE) {
        session_start();
    }
    $controller = new AuthController();
    $action = $_GET['action'] ?? 'loginRedirect';
    if (method_exists($controller, $action)) {
        $controller->$action();
    }
}
